python/lsst/afw/table/catalogMatches.py

```diff
@@ -0,0 +1,164 @@
+
+    if len(catalog) != len(target):
+        raise RuntimeError("Length mismatch: %d vs %d" % (len(catalog), len(target)))
+
+    sourceKeys = []
+    targetKeys = []
+    for k in sourceSchema.getNames():
+        if sourcePrefix is not None:
+            if not k.startswith(sourcePrefix):
+                continue
+            kFix = k[len(sourcePrefix):]
+        else:
+            kFix = k
+        sourceKeys.append(sourceSchema.find(k).key)
+        targetKeys.append(targetSchema.find((targetPrefix if targetPrefix is not None else "") + kFix).key)
+
+    for rFrom, rTo in zip(catalog, target):
+        for kFrom, kTo in zip(sourceKeys, targetKeys):
+            try:
+                rTo.set(kTo, rFrom.get(kFrom))
+            except:
+                print "Error setting: %s %s %s %s" % (type(rFrom), type(rTo), type(kFrom), type(kTo))
+
+    return target
+
+def matchesToCatalog(matches, matchMeta):
+    """Denormalise matches into a Catalog of "unpacked matches"
+
+    \param[in] matches    unpacked matches
+    \param[in] matchMeta  metadata for matches (must have .add attribute)
+
+    \return  lsst.afw.table.BaseCatalog of matches (with ref_ and src_ prefix identifiers
+             for referece and source entries, respectively)
+    """
+
+    if len(matches) == 0:
+        raise RuntimeError("No matches provided.")
+
+    refSchema = matches[0].first.getSchema()
+    srcSchema = matches[0].second.getSchema()
+
+    mergedSchema = copySchema(refSchema, Schema(), targetPrefix="ref_")
+    mergedSchema = copySchema(srcSchema, mergedSchema, targetPrefix="src_")
+    distKey = mergedSchema.addField("distance", type=float, doc="Distance between ref and src")
+
+    mergedCatalog = BaseCatalog(mergedSchema)
+    copyCatalog([m.first for m in matches], mergedCatalog, sourceSchema=refSchema, targetPrefix="ref_")
+    copyCatalog([m.second for m in matches], mergedCatalog, sourceSchema=srcSchema, targetPrefix="src_")
+    for m, r in zip(matches, mergedCatalog):
+        r.set(distKey, m.distance)
+
+    # obtaining reference catalog name if one is setup
+    try:
+        catalogName = os.path.basename(getPackageDir("astrometry_net_data"))
+    except:
+        catalogName = 'NOT_SET'
+    matchMeta.add('REFCAT', catalogName)
+    mergedCatalog.getTable().setMetadata(matchMeta)
+
+    return mergedCatalog
+
+def matchesFromCatalog(catalog, sourceSlotConfig=None, prefix=""):
+    """Generate a list of ReferenceMatches from a Catalog of "unpacked matches"
+
+    \param[in] catalog           catalog of matches.  Must have schema where reference entries are
+                                 prefixed with "ref_" and source entries are prefixed with "src_"
+    \param[in] sourceSlotConfig  configuration for source slots (optional)
+    \param[in] prefix            prefix for slot config setup (optional)
+
+    \returns   lsst.afw.table.ReferenceMatch of matches
+    """
+    if catalog is None:
+        # There are none
+        return []
```

Comment:
Is this check necessary? Since `catalog` isn't an optional argument, the caller would have to explicitly call `matchesFromCatalog(None)` -- surely the responsibility for not doing that lies with the caller rather than the callee?